README.md
# Overview

FastDFS Java Client API may be copied only under the terms of the BSD license.

## Dependencies

> for maven pom.xml

```xml
	<dependency>
		<groupId>in.clouthink.repack</groupId>
		<artifactId>fastdfs-client-java</artifactId>
		<version>1.29.4</version>
	</dependency>
```

> for gradle build.gradle

```gradle
	compile "in.clouthink.repack:fastdfs-client-java:1.29.4"
```

## Configuration sample

Format .conf(.ini) and .properties are support


> fastdfs-client.properties

```properties

fastdfs.connect_timeout_in_seconds = 5
fastdfs.network_timeout_in_seconds = 30

fastdfs.charset = UTF-8

fastdfs.http_anti_steal_token = false
fastdfs.http_secret_key = FastDFS1234567890
fastdfs.http_tracker_http_port = 80

fastdfs.tracker_servers = 10.0.11.201:22122,10.0.11.202:22122,10.0.11.203:22122

```

> fdfs_client.conf

```properties
connect_timeout = 2
network_timeout = 30
charset = UTF-8
http.tracker_http_port = 8080
http.anti_steal_token = no
http.secret_key = FastDFS1234567890

tracker_server = 10.0.11.247:22122
tracker_server = 10.0.11.248:22122
tracker_server = 10.0.11.249:22122

```


* 注1：tracker_server指向您自己IP地址和端口，1-n个
* 注2：除了tracker_server，其它配置项都是可选的


## Load configuration sample

> Load by file .conf

```java
ClientGlobal.init("fdfs_client.conf");
ClientGlobal.init("config/fdfs_client.conf");
ClientGlobal.init("/opt/fdfs_client.conf");
ClientGlobal.init("C:\\Users\\James\\config\\fdfs_client.conf");
```    

> Load by file .properties

```java
ClientGlobal.initByProperties("fastdfs-client.properties");
ClientGlobal.initByProperties("config/fastdfs-client.properties");
ClientGlobal.initByProperties("/opt/fastdfs-client.properties");
ClientGlobal.initByProperties("C:\\Users\\James\\config\\fastdfs-client.properties");
```

> Load by java.util.Properties
    
```java    
Properties props = new Properties();
props.put(ClientGlobal.PROP_KEY_TRACKER_SERVERS, "10.0.11.101:22122,10.0.11.102:22122");
ClientGlobal.initByProperties(props);
```

> Load by programming
    
```java
String trackerServers = "10.0.11.101:22122,10.0.11.102:22122";
ClientGlobal.initByTrackers(trackerServers);
```

> Check loaded status
    
```java
System.out.println("ClientGlobal.configInfo(): " + ClientGlobal.configInfo());
```
    
Output：

	ClientGlobal.configInfo(): {
	  g_connect_timeout(ms) = 5000
	  g_network_timeout(ms) = 30000
	  g_charset = UTF-8
	  g_anti_steal_token = false
	  g_secret_key = FastDFS1234567890
	  g_tracker_http_port = 80
	  trackerServers = 10.0.11.101:22122,10.0.11.102:22122
	}



RELEASE-NOTES.txt
Version 1.29.4  2019-02-14
  * add NameValuePair#toString

Version 1.29.3  2019-02-13
  * logging fastdfs call response code

Version 1.29.1  2019-02-13
  * Add more logging while uploading file

Version 1.29.0  2019-02-13
  * Camel Case Naming

Version 1.28.0  2019-02-12
  * Update pom.xml
  * Update javadoc

Version 1.26  2017-04-17
  * IniFileReader use getResourceAsStream first
  * change charactor encoding to UTF-8

Version 1.25  2014-12-07
  * support connection stats since FastDFS Server v5.04

Version 1.24  2012-12-01
  * bug fixed: StorageClient.get_file_info support appender file and 
    slave file correctly

Version 1.23  2012-11-18
  * query_file_info changed, you must 
    upgrade your FastDFS server to V4.03 or higher version

Version 1.22  2012-10-06
  * group and storage stat both add 1 field, you must 
    upgrade your FastDFS server to V4.00 or higher version

Version 1.21  2012-06-25
  * add exception class MyException
  * class IniFileReader same as FastDHT java client

Version 1.20  2012-05-27
  * appender file support truncate and modify operation
  * storage stat add 6 fields, you must upgrade your
    FastDFS server to V3.08 or higher version

Version 1.19  2011-05-29
  * storage stat add field if_trunk_server, you must upgrade your
    FastDFS server to V3.00 or higher version
  * tracker or storage close no throw IOException (silence close)

Version 1.18  2011-01-29
  * storage stat support total_upload_bytes, success_upload_bytes, 
    total_download_bytes and success_download_bytes etc. 18 fields
    you must upgrade your FastDFS server to V2.08 or higher version
  * StorageClient add functions: get_file_info and query_file_info,
    StorageClient1 add functions: get_file_info1 and query_file_info1

Version 1.17  2011-01-09
  * slave file's prefix name can be empty,
    you must upgrade your FastDFS server to V2.07 or higher version
  * support list group stats and storage stats
  * support delete storage server
  * add monitor class: org.csource.fastdfs.test.Monitor
  
Version 1.16  2010-10-17
  * upload file can get available storage server list of the group, 
    you must upgrade your FastDFS server to V2.01 or higher version

Version 1.15  2010-08-07
  * protocol reponse command changed. you must upgrade your FastDFS server
    to V2.00 or higher version
  * upload file protocol changed, you must upgrade your FastDFS server
    to V2.00 or higher version

Version 1.14  2010-06-22
  * bug fixed: delete the new created local file when download_file to 
    local file fail
  * support connect timeout for connecting, the network timeout used to 
    read and write

Version 1.13  2010-05-30
  * org.csource.fastdfs.ProtoCommon add method genSlaveFilename
  * support ACTIVE_TEST command. you must upgrade your FastDFS server
    to V1.28 or higher version

Version 1.12  2009-11-27
  * support uploading slave file
  * file ext name change from 5 to 6, you must upgrade your FastDFS server
    to V1.23 or higher version

Version 1.11  2009-10-22
  * correct interface / class name from beginning with Dowload to Download
  * auto reconnect with reused connection when IOException ocurs
  * test classes move to package org.csource.fastdfs.test
  * add classes org.csource.fastdfs.UploadStream and 
    org.csource.fastdfs.DownloadStream for streaming writing or reading
    
Version 1.10  2009-10-04
  * support multi-tracker server group, you can make multi 
    org.csource.fastdfs.TrackerClient instances for each tracker server group
  
Version 1.9  2009-10-01
  * add load test class: org.csource.fastdfs.TestLoad
  * network IO performance enhancement
  * upload file support callback for sending file content

Version 1.8  2009-08-30
  * TrackerClient, in function getConnection, add synchronized to avoid conflict
  * support file URL with token 
  
Version 1.7  2009-08-01
  * support cmd TRACKER_PROTO_CMD_SERVICE_QUERY_FETCH_ALL: query all storage 
    servers, you must upgrade your FastDFS server to V1.19 or higher version
    
Version 1.6  2009-05-24
  * download file support offset and download bytes, you must upgrade your
    FastDFS server to V1.18 or higher version

Version 1.5  2009-02-14
  * client can specify group name when upload file, you must upgrade your
    FastDFS server to V1.16 or higher version
    
Version 1.4  2008-11-10
  * upload file support ext name, you must upgrade your
    FastDFS server to V1.12 or higher version
  * upload protocol add field store_path_index (1 byte)
  * add method TrackerClient.getUpdateStorage. to support getmeta, setmeta 
    and delete file
  * TrackerClient.getFetchStorage bug fixed

Version 1.3  2008-10-16
  * client download file support callback function(class)
  * add class "StorageClient1" and function "TrackerClient.getFetchStorage1" 
    to combine 2 fields: group name and filename to 1 field: file id
  * add anothor client test program: TestClient1.java to use file id

Version 1.2  2008-09-08
  * communication protocol changed to support large file exceed 2GB: 
    all integer field is 8 bytes big-endian, 
    you must upgrade your FastDFS server to V1.8 or higher version

Version 1.1  2008-08-10
  * upload file package remove the one pad byte field, you must upgrade your
    FastDFS server to V1.4 or higher version

Version 1.0  2008-08-03
  * first version


src/main/java/org/csource/fastdfs/ClientGlobal.java
/**
 * Copyright (C) 2008 Happy Fish / YuQing
 * <p>
 * FastDFS Java Client may be copied only under the terms of the GNU Lesser
 * General Public License (LGPL).
 * Please visit the FastDFS Home Page http://www.csource.org/ for more detail.
 **/

package org.csource.fastdfs;

import org.csource.common.IniFileReader;
import org.csource.common.FastdfsException;

import java.io.IOException;
import java.io.InputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

/**
 * Global variables
 *
 * @author Happy Fish / YuQing
 * @version Version 1.11
 */
public class ClientGlobal {

    public static final String CONF_KEY_CONNECT_TIMEOUT        = "connect_timeout";
    public static final String CONF_KEY_NETWORK_TIMEOUT        = "network_timeout";
    public static final String CONF_KEY_CHARSET                = "charset";
    public static final String CONF_KEY_HTTP_ANTI_STEAL_TOKEN  = "http.anti_steal_token";
    public static final String CONF_KEY_HTTP_SECRET_KEY        = "http.secret_key";
    public static final String CONF_KEY_HTTP_TRACKER_HTTP_PORT = "http.tracker_http_port";
    public static final String CONF_KEY_TRACKER_SERVER         = "tracker_server";

    public static final String PROP_KEY_CONNECT_TIMEOUT_IN_SECONDS = "fastdfs.connect_timeout_in_seconds";
    public static final String PROP_KEY_NETWORK_TIMEOUT_IN_SECONDS = "fastdfs.network_timeout_in_seconds";
    public static final String PROP_KEY_CHARSET                    = "fastdfs.charset";
    public static final String PROP_KEY_HTTP_ANTI_STEAL_TOKEN      = "fastdfs.http_anti_steal_token";
    public static final String PROP_KEY_HTTP_SECRET_KEY            = "fastdfs.http_secret_key";
    public static final String PROP_KEY_HTTP_TRACKER_HTTP_PORT     = "fastdfs.http_tracker_http_port";
    public static final String PROP_KEY_TRACKER_SERVERS            = "fastdfs.tracker_servers";

    public static final int     DEFAULT_CONNECT_TIMEOUT        = 5; //second
    public static final int     DEFAULT_NETWORK_TIMEOUT        = 30; //second
    public static final String  DEFAULT_CHARSET                = "UTF-8";
    public static final boolean DEFAULT_HTTP_ANTI_STEAL_TOKEN  = false;
    public static final String  DEFAULT_HTTP_SECRET_KEY        = "FastDFS1234567890";
    public static final int     DEFAULT_HTTP_TRACKER_HTTP_PORT = 80;

    public static int     connectTimeout  = DEFAULT_CONNECT_TIMEOUT * 1000; //millisecond
    public static int     networkTimeout  = DEFAULT_NETWORK_TIMEOUT * 1000; //millisecond
    public static String  charset         = DEFAULT_CHARSET;
    public static boolean antiStealToken  = DEFAULT_HTTP_ANTI_STEAL_TOKEN; //if anti-steal token
    public static String  secretKey       = DEFAULT_HTTP_SECRET_KEY; //generage token secret key
    public static int     trackerHttpPort = DEFAULT_HTTP_TRACKER_HTTP_PORT;

    public static TrackerGroup trackerGroup;

    private ClientGlobal() {
    }

    /**
     * load global variables
     *
     * @param conf_filename config filename
     * @throws IOException ioex
     * @throws FastdfsException fastex
     */
    public static void init(String conf_filename) throws IOException, FastdfsException {
        IniFileReader iniReader;
        String[] szTrackerServers;
        String[] parts;

        iniReader = new IniFileReader(conf_filename);

        connectTimeout = iniReader.getIntValue("connect_timeout", DEFAULT_CONNECT_TIMEOUT);
        if (connectTimeout < 0) {
            connectTimeout = DEFAULT_CONNECT_TIMEOUT;
        }
        connectTimeout *= 1000; //millisecond

        networkTimeout = iniReader.getIntValue("network_timeout", DEFAULT_NETWORK_TIMEOUT);
        if (networkTimeout < 0) {
            networkTimeout = DEFAULT_NETWORK_TIMEOUT;
        }
        networkTimeout *= 1000; //millisecond

        charset = iniReader.getStrValue("charset");
        if (charset == null || charset.length() == 0) {
            charset = "ISO8859-1";
        }

        szTrackerServers = iniReader.getValues("tracker_server");
        if (szTrackerServers == null) {
            throw new FastdfsException("item \"tracker_server\" in " + conf_filename + " not found");
        }

        InetSocketAddress[] tracker_servers = new InetSocketAddress[szTrackerServers.length];
        for (int i = 0; i < szTrackerServers.length; i++) {
            parts = szTrackerServers[i].split("\\:", 2);
            if (parts.length != 2) {
                throw new FastdfsException("the value of item \"tracker_server\" is invalid, the correct format is host:port");
            }

            tracker_servers[i] = new InetSocketAddress(parts[0].trim(), Integer.parseInt(parts[1].trim()));
        }
        trackerGroup = new TrackerGroup(tracker_servers);

        trackerHttpPort = iniReader.getIntValue("http.tracker_http_port", 80);
        antiStealToken = iniReader.getBoolValue("http.anti_steal_token", false);
        if (antiStealToken) {
            secretKey = iniReader.getStrValue("http.secret_key");
        }
    }

    /**
     * load from properties file
     *
     * @param propsFilePath properties file path, eg:
     *                      "fastdfs-client.properties"
     *                      "config/fastdfs-client.properties"
     *                      "/opt/fastdfs-client.properties"
     *                      "C:\\Users\\James\\config\\fastdfs-client.properties"
     *                      properties文件至少包含一个配置项 fastdfs.tracker_servers 例如：
     *                      fastdfs.tracker_servers = 10.0.11.245:22122,10.0.11.246:22122
     *                      server的IP和端口用冒号':'分隔
     *                      server之间用逗号','分隔
     * @throws IOException ioex
     * @throws FastdfsException fastex
     */
    public static void initByProperties(String propsFilePath) throws IOException, FastdfsException {
        Properties props = new Properties();
        InputStream in = IniFileReader.loadFromOsFileSystemOrClasspathAsStream(propsFilePath);
        if (in != null) {
            props.load(in);
        }
        initByProperties(props);
    }

    public static void initByProperties(Properties props) throws IOException, FastdfsException {
        String trackerServersConf = props.getProperty(PROP_KEY_TRACKER_SERVERS);
        if (trackerServersConf == null || trackerServersConf.trim().length() == 0) {
            throw new FastdfsException(String.format("configure item %s is required", PROP_KEY_TRACKER_SERVERS));
        }
        initByTrackers(trackerServersConf.trim());

        String connectTimeoutInSecondsConf = props.getProperty(PROP_KEY_CONNECT_TIMEOUT_IN_SECONDS);
        String networkTimeoutInSecondsConf = props.getProperty(PROP_KEY_NETWORK_TIMEOUT_IN_SECONDS);
        String charsetConf = props.getProperty(PROP_KEY_CHARSET);
        String httpAntiStealTokenConf = props.getProperty(PROP_KEY_HTTP_ANTI_STEAL_TOKEN);
        String httpSecretKeyConf = props.getProperty(PROP_KEY_HTTP_SECRET_KEY);
        String httpTrackerHttpPortConf = props.getProperty(PROP_KEY_HTTP_TRACKER_HTTP_PORT);
        if (connectTimeoutInSecondsConf != null && connectTimeoutInSecondsConf.trim().length() != 0) {
            connectTimeout = Integer.parseInt(connectTimeoutInSecondsConf.trim()) * 1000;
        }
        if (networkTimeoutInSecondsConf != null && networkTimeoutInSecondsConf.trim().length() != 0) {
            networkTimeout = Integer.parseInt(networkTimeoutInSecondsConf.trim()) * 1000;
        }
        if (charsetConf != null && charsetConf.trim().length() != 0) {
            charset = charsetConf.trim();
        }
        if (httpAntiStealTokenConf != null && httpAntiStealTokenConf.trim().length() != 0) {
            antiStealToken = Boolean.parseBoolean(httpAntiStealTokenConf);
        }
        if (httpSecretKeyConf != null && httpSecretKeyConf.trim().length() != 0) {
            secretKey = httpSecretKeyConf.trim();
        }
        if (httpTrackerHttpPortConf != null && httpTrackerHttpPortConf.trim().length() != 0) {
            trackerHttpPort = Integer.parseInt(httpTrackerHttpPortConf);
        }
    }

    /**
     * load from properties file
     *
     * @param trackerServers 例如："10.0.11.245:22122,10.0.11.246:22122"
     *                       server的IP和端口用冒号':'分隔
     *                       server之间用逗号','分隔
     * @throws IOException ioex
     * @throws FastdfsException fastex
     */
    public static void initByTrackers(String trackerServers) throws IOException, FastdfsException {
        List<InetSocketAddress> list = new ArrayList();
        String spr1 = ",";
        String spr2 = ":";
        String[] arr1 = trackerServers.trim().split(spr1);
        for (String addrStr : arr1) {
            String[] arr2 = addrStr.trim().split(spr2);
            String host = arr2[0].trim();
            int port = Integer.parseInt(arr2[1].trim());
            list.add(new InetSocketAddress(host, port));
        }
        InetSocketAddress[] trackerAddresses = list.toArray(new InetSocketAddress[list.size()]);
        initByTrackers(trackerAddresses);
    }

    public static void initByTrackers(InetSocketAddress[] trackerAddresses) throws IOException, FastdfsException {
        trackerGroup = new TrackerGroup(trackerAddresses);
    }

    /**
     * construct Socket object
     *
     * @param ip_addr ip address or hostname
     * @param port    port number
     * @return connected Socket object
     * @throws IOException ioex
     */
    public static Socket getSocket(String ip_addr, int port) throws IOException {
        Socket sock = new Socket();
        sock.setSoTimeout(ClientGlobal.networkTimeout);
        sock.connect(new InetSocketAddress(ip_addr, port), ClientGlobal.connectTimeout);
        return sock;
    }

    /**
     * construct Socket object
     *
     * @param addr InetSocketAddress object, including ip address and port
     * @return connected Socket object
     * @throws IOException ioex
     */
    public static Socket getSocket(InetSocketAddress addr) throws IOException {
        Socket sock = new Socket();
        sock.setSoTimeout(ClientGlobal.networkTimeout);
        sock.connect(addr, ClientGlobal.connectTimeout);
        return sock;
    }

    public static int getConnectTimeout() {
        return connectTimeout;
    }

    public static void setConnectTimeout(int connect_timeout) {
        ClientGlobal.connectTimeout = connect_timeout;
    }

    public static int getNetworkTimeout() {
        return networkTimeout;
    }

    public static void setNetworkTimeout(int network_timeout) {
        ClientGlobal.networkTimeout = network_timeout;
    }

    public static String getCharset() {
        return charset;
    }

    public static void setCharset(String charset) {
        ClientGlobal.charset = charset;
    }

    public static int getTrackerHttpPort() {
        return trackerHttpPort;
    }

    public static void setTrackerHttpPort(int tracker_http_port) {
        ClientGlobal.trackerHttpPort = tracker_http_port;
    }

    public static boolean getAntiStealToken() {
        return antiStealToken;
    }

    public static boolean isAntiStealToken() {
        return antiStealToken;
    }

    public static void setAntiStealToken(boolean anti_steal_token) {
        ClientGlobal.antiStealToken = anti_steal_token;
    }

    public static String getSecretKey() {
        return secretKey;
    }

    public static void setSecretKey(String secret_key) {
        ClientGlobal.secretKey = secret_key;
    }

    public static TrackerGroup getTrackerGroup() {
        return trackerGroup;
    }

    public static void setTrackerGroup(TrackerGroup tracker_group) {
        ClientGlobal.trackerGroup = tracker_group;
    }

    public static String configInfo() {
        String trackerServers = "";
        if (trackerGroup != null) {
            InetSocketAddress[] trackerAddresses = trackerGroup.tracker_servers;
            for (InetSocketAddress inetSocketAddress : trackerAddresses) {
                if (trackerServers.length() > 0) {
                    trackerServers += ",";
                }
                trackerServers += inetSocketAddress.toString().substring(1);
            }
        }
        return "{" + "\n  connect_timeout(ms) = " + connectTimeout + "\n  network_timeout(ms) = " + networkTimeout + "\n  charset = " +
                charset + "\n  anti_steal_token = " + antiStealToken +
                "\n  secret_key = " + secretKey + "\n  tracker_http_port = " + trackerHttpPort +
                "\n  trackerServers = " + trackerServers + "\n}";
    }

}
